rfcs/20220309-feature_flags.md

```diff
@@ -0,0 +1,229 @@
+
+Due to their nature being run-time rather than build-time, they are best used in pairing with another storage (as an override for the default) rather than on their own.
+
+For example, we can have a feature flag which is set to disabled via a config file. This prevents users from accessing this feature. However, a developer who still wants to use this feature can send the query param as true and access it.
+
+#### 3. Growthbook
+
+GrowthBook is an open-source platform for feature flagging. It provides SDKs and a UI to manage feature flags.
+
+##### Pros
+
+1. High-level solution with UI.
+1. Supports features we'd want later like A/B testing.
+1. Can record a large number of metrics.
+1. Actively maintained and FOSS.
+1. No redeployment necessary, flags can be toggled remotely.
+1. Allows flags to have many sub-attributes.
+
+##### Cons
+
+1. Adds API calls to the server startup.
+1. Separate service that needs to be self-hosted and set up.
+1. Will need Dockerisation of the dev workflow.
+1. Might be overkill if we don't want the extra features.
+1. We might not even want to use the metrics.
+
+##### Reference
+
+- [Home](https://www.growthbook.io)
+- [Docs](https://docs.growthbook.io)
+
+#### 4. API
+
+Our API could serve as a decent source of feature flag statuses by storing them as Django model.
+
+##### Pros
+
+1. Can be managed using Django admin.
+1. No redeployment necessary, flags can be toggled remotely.
+1. Allows flags to have many sub-attributes, essentially models.
+1. The frontend depends on the API anyway, so just add more of it.
+
+##### Cons
+
+1. Need to mock the API calls in dev environments.
+1. Adds coupling between frontend features and the API.
+
+#### 5. Conf files
+
+This is functionally similar to the evironment variables storage option, but has many advantages.
+
+Conf files could be JSON/YAML files that list the names of features and their values. They can live separate from the codebase, athough there's really no advantage or disadvantage to them being with or separate from the code.
+
+##### Pros
+
+1. Simple configuration.
+1. Prevents lots of API calls.
+1. Can store complex configurations and nesting, which environment variables cannot.
+1. Easier to manage being persistent files rather than machine/container state.
+
+##### Cons
+
+1. YAML is easier to write by hand but needs a separate library to read.
+1. JSON is easier to read but is not designed for writing manually.
+
+Regardless of the storage mechanism we will need a few other key things.
+
+### Store
+
+To prevent feature flags from being all over the place, we can keep them store in a Pinia store and just use the store wherever there is a check. This allows us to toggle features from a centralised place.
+
+Since keys will be added and removed periodically, the store must not expose feature flags directly but rather expose a getter that can handle fallback-to-default.
+
+The flags can be populated in the store from whatever source we choose.
```

Comment:
This can be tricky. Pinia stores are created when you call `useXStore`, so they are not necessarily called on the server, or when Nuxt is being built.
Where would you be calling the flags store to create it and populate it from the source?
If we use `middleware`, then the reading would be done for every request.

Reply:
>  If we use middleware, then the reading would be done for every request.

If we use a JSON file we can just `import` it like a regular JS module and the "reading" will be done once and only once. There would be no performance impact in that approach.